### PR TITLE
CMC-388: Callback handler framework

### PR DESCRIPTION
**Pull request by sabahirfan**
### JIRA link (if applicable) ###
[CMC-388](https://tools.hmcts.net/jira/browse/CMC-388)

### Change description ###
This PR implements an abstract factory framework for callbacks to reduce the boiler point code.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package uk.gov.hmcts.reform.ucmc.callback;
+
+import uk.gov.hmcts.reform.ccd.client.model.CallbackResponse;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public abstract class CallbackHandler {
+
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java:9]: remove extra line

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    public abstract List<CaseEvent> handledEvents();
+
+    public void register(Map<String, CallbackHandler> handlers) {
+        handledEvents().forEach(
+            handledEvent -> handlers.put(handledEvent.getValue(), this));
+    }
+
+    public CallbackResponse handle(CallbackParams callbackParams) {
+        return Optional.ofNullable(callbacks().get(callbackParams.getType()))
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java]: Optional.ofNullable could be static import

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandlerFactory.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        beans.forEach(bean -> bean.register(eventHandlers));
+    }
+
+    public CallbackResponse dispatch(CallbackParams callbackParams) {
+        String eventId = callbackParams.getRequest().getEventId();
+        return Optional.ofNullable(eventHandlers.get(eventId))
+            .map(h -> h.handle(callbackParams))
+            .orElseThrow(() -> new CallbackException("Could not handle callback for event " + eventId));
+    }
+
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandlerFactory.java]: can remove line

File: src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @PathVariable("callback-type") String callbackType,
+        @NotNull @RequestBody CallbackRequest callback,
+        @PathVariable("version") Optional<String> version
+    ) {
+        log.info("Received callback from CCD, eventId: {}", callback.getEventId());
+        CallbackParams callbackParams = CallbackParams.builder()
+            .request(callback)
+            .type(CallbackType.fromValue(callbackType))
+            .params(ImmutableMap.of(CallbackParams.Params.BEARER_TOKEN, authorisation))
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java:55]: could do Map.of()

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+@ExtendWith(MockitoExtension.class)
+public class CallbackHandlerFactoryTest {
+
+    public static final String BEARER_TOKEN = "Bearer Token";
+    public static final CallbackResponse response = AboutToStartOrSubmitCallbackResponse.builder().build();
+
+    private CallbackHandler sampleCallbackHandler = new CallbackHandler() {
+        @Override
+        protected Map<CallbackType, Callback> callbacks() {
+            return ImmutableMap.of(
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: could use Map.of()

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        public List<CaseEvent> handledEvents() {
+            return Collections.singletonList(CREATE_CASE);
+        }
+    };
+
+    private CallbackHandlerFactory callbackHandlerFactory;
+
+    @BeforeEach
+    public void setUp() {
+        callbackHandlerFactory = new CallbackHandlerFactory(ImmutableList.of(sampleCallbackHandler));
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: could use List.of()

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    public void shouldThrowIfUnknownEvent() {
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: thoughts on test names following should ... when ... ?

shouldThrowCallbackExceptionWhenUnknownEvent?

File: build.gradle [hunk 8]
```diff
@@ -1,7 +1,6 @@
 plugins {
   id 'application'
   id 'checkstyle'
-  id 'pmd'
```
Comment (pd-dudkiewicz) [inline build.gradle:4]: Do we really want to remove pmd? We could leave it and maybe configure ruleset to match our needs.

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackException.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmcts.reform.ucmc.callback;
+
+public class CallbackException extends RuntimeException {
+    public CallbackException(String message) {
+        super(message);
+    }
+
+    public CallbackException(String message, Throwable t) {
```
Comment (pd-dudkiewicz) [inline src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackException.java]: Shall we remove it if it's not used?

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            .request(callbackRequest)
+            .params(ImmutableMap.of(CallbackParams.Params.BEARER_TOKEN, BEARER_TOKEN))
+            .build();
+
+        Exception exception = assertThrows(
+            CallbackException.class,
+            () -> callbackHandlerFactory.dispatch(params)
+        );
+
+        assertEquals("Could not handle callback for event nope", exception.getMessage());
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: this could use assertThatThrownBy assertion so that 

```
assertThatThrownBy(() -> callbackHandlerFactory.dispatch(params))
            .isInstanceOf(CallbackException.class)
            .hasMessage("Could not handle callback for event nope");
```

File: src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        this.callbackHandlerFactory = callbackHandlerFactory;
+    }
+
+    @PostMapping(path = {"/{callback-type}", "{version}/{callback-type}"})
+    @ApiOperation("Handles all callbacks from CCD")
+    public CallbackResponse callback(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @PathVariable("callback-type") String callbackType,
+        @NotNull @RequestBody CallbackRequest callback,
+        @PathVariable("version") Optional<String> version
```
Comment (pd-dudkiewicz) [inline src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java]: Could `CallbackVersion` enum be used here?

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static uk.gov.hmcts.reform.ucmc.callback.CallbackType.ABOUT_TO_SUBMIT;
+import static uk.gov.hmcts.reform.ucmc.callback.CaseEvent.CREATE_CASE;
+
+@ExtendWith(MockitoExtension.class)
+public class CallbackHandlerFactoryTest {
+
+    public static final String BEARER_TOKEN = "Bearer Token";
+    public static final CallbackResponse response = AboutToStartOrSubmitCallbackResponse.builder().build();
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: is there any value in this being defined here? I see it only used one in the second test

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+            .request(callbackRequest)
+            .type(ABOUT_TO_SUBMIT)
+            .params(ImmutableMap.of(CallbackParams.Params.BEARER_TOKEN, BEARER_TOKEN))
+            .build();
+
+        CallbackResponse callbackResponse = callbackHandlerFactory.dispatch(params);
+
+        assertEquals(response, callbackResponse);
+    }
+
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: can remove line

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        Exception exception = assertThrows(
+            CallbackException.class,
+            () -> callbackHandlerFactory.dispatch(params)
+        );
+
+        assertEquals("Could not handle callback for event nope", exception.getMessage());
+    }
+
+    @Test
+    public void shouldDispatchCallbackForEvent() {
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: maybe look at test name, depending on previous comment

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackTypeTest.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Test
+    public void shouldDeserialiseValidCallbacks() {
+        assertThat(CallbackType.fromValue("mid")).isEqualTo(CallbackType.MID);
+        assertThat(CallbackType.fromValue("about-to-start"))
+            .isEqualTo(CallbackType.ABOUT_TO_START);
+        assertThat(CallbackType.fromValue("about-to-submit"))
+            .isEqualTo(CallbackType.ABOUT_TO_SUBMIT);
+        assertThat(CallbackType.fromValue("submitted"))
+            .isEqualTo(CallbackType.SUBMITTED);
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackTypeTest.java]: This could be a parameterized test or could extract a common is equal method. But not very important

**Review by swalker125 — DISMISSED** [review]
Couple of things, but functionally I have no issues

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java (hunk 12 above)
Comment (pd-dudkiewicz) [inline src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: The name should be in uppercase as it's constant.

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java (hunk 7 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: done

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java (hunk 12 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: This is being used in sampleClaimHandler class we well.

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java (hunk 14 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: done

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java (hunk 12 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: done

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackException.java (hunk 9 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackException.java]: removed

File: src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java (hunk 11 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java]: agree that will be better.

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandlerFactory.java (hunk 3 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandlerFactory.java]: removed

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java (hunk 10 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: done

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java (hunk 1 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java:9]: removed.

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java (hunk 2 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java]: done

File: src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java (hunk 13 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/callback/CallbackHandlerFactoryTest.java]: removed

File: src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java (hunk 4 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java:55]: Prefer to use Immutable collections as they are fully immutable, nulls are forbidden and (slightly) more time- and space-efficient compared to usual Java's collections. Happy to hear your thoughts.

File: build.gradle (hunk 8 above)
Comment (sabahirfan) [reply build.gradle:4]: We are only using checkstyle in most of the reform projects, as there is overlap between both of these tools. I also saw this being removed in one of our existing PR in this project, so thought we are getting rid of it. Please advise otherwise.

File: src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java (hunk 4 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/controllers/CallbackController.java:55]: Resolving as discussed over the chat.

File: build.gradle (hunk 8 above)
Comment (pd-dudkiewicz) [reply build.gradle:4]: I don't really have a strong opinion on it, if it doesn't provide much value then feel free to remove it. We didn't use it in previous project either. Resolving the conversation.

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.ResultActions;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+import uk.gov.hmcts.reform.ucmc.Application;
+
+@SpringBootTest(
+    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
```
Comment (pd-dudkiewicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: what's the value of setting random port?

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    protected static final String BEARER_TOKEN = "Bearer let me in";
+
+    @Autowired
+    protected ObjectMapper objectMapper;
+
+    @Autowired
+    protected MockMvc mockMvc;
+
+    protected ResultActions doPost(String auth, String content, String urlTemplate, Object... uriVars)
+        throws Exception {
```
Comment (pd-dudkiewicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: could we handle the exception here, so we don't have to add `throws Exception` to every test (it's unnecessary noise imo)?

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+    protected static final String BEARER_TOKEN = "Bearer let me in";
+
+    @Autowired
+    protected ObjectMapper objectMapper;
+
+    @Autowired
+    protected MockMvc mockMvc;
+
+    protected ResultActions doPost(String auth, String content, String urlTemplate, Object... uriVars)
```
Comment (pd-dudkiewicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: how about `content` being CallbackRequest, so we don't have to wrap it in `toJson()` in every subclass (it could be done inside this method)?

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.test.web.servlet.ResultActions;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+import uk.gov.hmcts.reform.ucmc.Application;
+
+@SpringBootTest(
+    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
+    classes = Application.class
+)
+@AutoConfigureMockMvc
+public class BaseIntegrationTest {
```
Comment (pd-dudkiewicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: should this be abstract class?

File: src/main/resources/application.yaml [hunk 20]
```diff
@@ -15,3 +15,5 @@ management:
 spring:
   application:
     name: ucmc-service
+  main:
+    allow-bean-definition-overriding: true
```
Comment (pd-dudkiewicz) [inline src/main/resources/application.yaml]: this doesn't feel right, what's the reason behind this change?

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java (hunk 18 above)
Comment (sabahirfan) [reply src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: As this is base class, we may want to pass some other objects as well for different integration tests. I think generic type would be better here.

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java (hunk 19 above)
Comment (sabahirfan) [reply src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: yes. done

File: src/main/resources/application.yaml (hunk 20 above)
Comment (sabahirfan) [reply src/main/resources/application.yaml]: Preview was complaining about this bean. `core-case-data-api.FeignClientSpecification`. We have to do similar change in claim-store since sprintboot version 2.1. Do we have alternate solution? Would be happy to adopt that.

```  
OverrideException: Invalid bean definition with name 'core-case-data-api.FeignClientSpecification' defined in null: Cannot register bean definition [Generic bean: class [org.springframework.cloud.openfeign.FeignClientSpecification]; scope=; abstract=false; lazyInit=null; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=null; factoryMethodName=null; initMethodName=null; destroyMethodName=null] for bean 'core-case-data-api.FeignClientSpecification': There is already [Generic bean: class [org.springframework.cloud.openfeign.FeignClientSpecification]; scope=; abstract=false; lazyInit=null; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=null; factoryMethodName=null; initMethodName=null; destroyMethodName=null] bound.
2020-07-07T08:39:39.983 INFO  [main] o.s.b.a.l.ConditionEvaluationReportLoggingListener
```

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java (hunk 17 above)
Comment (sabahirfan) [reply src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: There would be other reasons as well to throw Exception in test methods i.e. `ResultActions.andExpect()` or `ResultActions.andDo(` etc. If we are ok to use `@SneakyThrow` from lombok in our tests only, we can use that to reduce this noise.

File: build.gradle [hunk 21]
```diff
@@ -76,6 +75,7 @@ task integration(type: Test) {
   testClassesDirs = sourceSets.integrationTest.output.classesDirs
   classpath = sourceSets.integrationTest.runtimeClasspath
   failFast = true
+  testLogging.showStandardStreams = false
```
Comment (dharmendrak) [inline build.gradle:79]: this can be removed now

**Review by dharmendrak — DISMISSED** [review]
LGTM

File: build.gradle (hunk 21 above)
Comment (sabahirfan) [reply build.gradle:79]: Prefer to have it there as it's easy to turn flag on locally to see the logs, rather adding the whole property when needed.

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java (hunk 16 above)
Comment (sabahirfan) [reply src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: removed.

File: src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java (hunk 17 above)
Comment (sabahirfan) [reply src/integrationTest/java/uk/gov/hmcts/reform/ucmc/controllers/BaseIntegrationTest.java]: As discussed replacing this with @SneakyThrows.

**Review by dharmendrak — DISMISSED** [review]

File: src/main/resources/application.yaml (hunk 20 above)
Comment (sabahirfan) [reply src/main/resources/application.yaml]: https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.1-Release-Notes#bean-overriding

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+        handledEvents().forEach(
+            handledEvent -> handlers.put(handledEvent.getValue(), this));
+    }
+
+    public CallbackResponse handle(CallbackParams callbackParams) {
+        return ofNullable(callbacks().get(callbackParams.getType()))
+            .map(callback -> callback.execute(callbackParams))
+            .orElseThrow(() -> new CallbackException(
+                String.format(
+                    "Callback for event %s, type %s not implemented",
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java:26]: do we want to restructure this message to read something like "{type of callback} callback for event {event} not implemented"

Not super important

File: src/main/resources/application.yaml (hunk 20 above)
Comment (swalker125) [reply src/main/resources/application.yaml]: someone's been copy and pasting 😛 
https://github.com/hmcts/ccd-client/blob/master/src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java#L27
https://github.com/hmcts/ccd-client/blob/master/src/main/java/uk/gov/hmcts/reform/ccd/client/CaseEventsApi.java#L17

shall we fix this in ccd-client rather than accepting the error?

File: src/test/java/uk/gov/hmcts/reform/ucmc/advice/ResourceExceptionHandlerTest.java [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+    private <E extends Exception> void testTemplate(
+        String message,
+        Function<String, E> exceptionBuilder,
+        Function<E, ResponseEntity<?>> method,
+        HttpStatus expectedStatus
+    ) {
+        E exception = exceptionBuilder.apply(message);
+        ResponseEntity<?> result = method.apply(exception);
+        assertThat(result.getStatusCode()).isSameAs(expectedStatus);
+        assertThat(result.getBody()).isNotNull()
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/ucmc/advice/ResourceExceptionHandlerTest.java:41]: is the isNotNull necessary?

**Review by swalker125 — CHANGES_REQUESTED** [review]
not a fan of bean override!

File: src/main/resources/application.yaml (hunk 20 above)
Comment (pd-dudkiewicz) [reply src/main/resources/application.yaml]: Well spotted @swalker125, that's why we should avoid bean overriding, it usually hides the real problem. In the meanwhile we could go for previous `4.6.4` version of ccd-client

Comment (sabahirfan) [reply src/main/resources/application.yaml]: https://github.com/hmcts/ccd-client/pull/60

File: src/test/java/uk/gov/hmcts/reform/ucmc/advice/ResourceExceptionHandlerTest.java (hunk 23 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/ucmc/advice/ResourceExceptionHandlerTest.java:41]: yes ResponseEntity is sending the message in body.

File: src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java (hunk 22 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/ucmc/callback/CallbackHandler.java:26]: I will pass this :)

**Review by swalker125 — APPROVED** [review]

**Review by dharmendrak — APPROVED** [review]
